keras/engine/training.py

```diff
@@ -258,7 +259,16 @@ def collect_trainable_weights(layer):
         weights += layer.trainable_weights
     # dedupe weights
     weights = list(set(weights))
-    weights.sort(key=lambda x: x.name)
+    # TF variables have auto-generated the name, while Theano has auto-generated the auto_name variable. name in Theano is None
+    if weights:
+        if K._BACKEND == 'tensorflow':
+            weights.sort(key=lambda x: x.name)
+        elif K._BACKEND == 'theano':
+            # auto_name looks e.g. like this 'auto_11', so we have to parse the int and sort by that
```

Comment:
Actually we don't; what matters is that the weights are in a deterministic order. In *what* order they are is not very important. We could simply sort by `auto_name`.